Migrated Profiles to RQ

mzparacha opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Link to Issue
Closes: #4892, #4819

## Description of Changes
- Migrated `newProfile` controller APIs to react query and cleaned up unnecessary state/apis and logic
- Cleaned up `user.tsx` to be more descriptive and removed unnecessary logic

## "How We Fixed It"
<!-- Brief description of solution, if bug, to be added once issue is resolved. -->
Not a bug, just a migration. We migrated `newProfile` controller to react query.

## Test Plan
- Visit different pages that show user info (essentially the user.tsx) component and validate that correct data is displayed for each user
- Go to edit profile and make sure that editing your profile works perfectly and visiting subsequent pages after that work perfectly as well (without reloading the page)
- [Important] - Make sure that creating a new account works perfectly with multiple wallets
- [Important] - Make sure that logging into an existing account works perfectly and visiting subsequent pages after that work perfectly as well (without reloading the page)

## Deployment Plan
N/A

## Other Considerations
- The `newProfile` is not entirely cleaned up. I left some event emitters there intentionally as to not break the wallet/auth logic. We need to come up with plan to cleanup manual event emitters. @jnaviask @masvelio 
- The `newProfile` tests would be broken after this change, we can remove those tests.

mzparacha commented on packages/commonwealth/client/scripts/controllers/server/newProfiles.ts:
The remaining code in this file is left over on purpose. We need to come up with a plan to remove these event emitters as some wallet code depended on it and it can break very easily on some emitters change.

masvelio commented on packages/commonwealth/client/scripts/views/components/Header/LoginSelector.tsx:
```diff
@@ -90,7 +90,7 @@ export const LoginSelector = () => {
   }
 
-  if (!profileLoadComplete && NewProfilesController.Instance.allLoaded()) {
+  if (!profileLoadComplete) {
     setProfileLoadComplete(true);
   }
```
1. is this even necessary here? If I removed this `if` statement, login selector seems to work fine? If not necessary, this code below could be removed
```
  useEffect(() => {
    NewProfilesController.Instance.isFetched.on('redraw', () => {
      setProfileLoadComplete(true);
    });
  }, []);
  ```
2. If this is necessary though, I know this is not your code, but setting state should not be in render (it should be in useEffect)

mzparacha replied:
Hm, `NewProfilesController.Instance.isFetched.on(` actually sets up a listener, so the state will update on the listener callback. 

TBH, I don't understand why `if (!profileLoadComplete) {` was added originally but it seem to sometimes break the login on metamask in my tests. We can do more digging here I suppose, will do more and update...

masvelio replied:
I know that it sets up a listener but the question is if we need this listener here? As I said, I removed it along with the setProfileLoadComplete and I did not see any errors. User Avatar has loaded properly. Please test it around and remove if everuthing works. We want eventually get rid of  `NewProfilesController.Instance.isFetched` anyways

masvelio commented on packages/commonwealth/client/scripts/views/components/ReactionButton/helpers.tsx:
```diff
@@ -24,8 +22,9 @@ export const getDisplayedReactorsForPopup = ({
             <User
-              user={new AddressInfo(null, rxn, app.chain.id, null)}
-              linkify
+              userAddress={rxn}
```
what `rxn` means? Let's be more descriptive

masvelio commented on packages/commonwealth/client/scripts/views/components/component_kit/cw_sidebar_menu.tsx:
```diff
@@ -2,18 +2,17 @@ import React, { useState } from 'react';
 import 'components/component_kit/cw_sidebar_menu.scss';
 
+import { featureFlags } from 'helpers/feature-flags';
+import { navigateToCommunity, useCommonNavigate } from 'navigation/helpers';
 import app from 'state';
-import AddressInfo from '../../../models/AddressInfo';
+import useSidebarStore from 'state/ui/sidebar';
```
I can see some resorting in imports in this PR. Is this some automation or you do it manually?

mzparacha replied:
`Command + Shift + O` in vscode 😄, orders it automatically. Please lmk if i should revert.

mzparacha replied:
Actually its using this extension by default https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint&ssr=false 🤔.

I think we can set up a lint config that works cross IDEs. (not sure if its just my IDE alone that's messing up 😅)

mzparacha replied:
[Done](https://github.com/hicommonwealth/commonwealth/pull/4865/commits/94a01227f5dd0f1b6f2f968eb1d5a6d4053e2efd)

mzparacha commented on packages/commonwealth/client/scripts/hooks/useWallets.tsx:
```diff
@@ -367,11 +371,20 @@ const useWallets = (walletProps: IuseWalletProps) => {
+      await DISCOURAGED_NONREACTIVE_getProfilesByAddress(
+        primaryAccount.profile.chain,
+        primaryAccount.profile.address
+      ).then((res) => {
```
hm, that would work as well. Will update shortly.

mzparacha replied:
Done

mzparacha commented on packages/commonwealth/client/scripts/hooks/useWallets.tsx:
```diff
@@ -416,16 +429,16 @@ const useWallets = (walletProps: IuseWalletProps) => {
+          address: primaryAccount.profile.address,
+          name: username,
+          avatarUrl,
+        }).then(() => {
+          // we should trigger a redraw emit manually
+          NewProfilesController.Instance.isFetched.emit('redraw');
+        });
```
yes, we don't need it. I don't remember why I kept doing `.then` 😅

mzparacha replied:
Done

mzparacha commented on packages/commonwealth/client/scripts/state/api/profiles/banProfileByAddress.ts:
```diff
@@ -0,0 +1,33 @@
+const useBanProfileByAddressMutation = ({
+  chainId,
+  address,
+}: BanProfileByAddressProps) => {
```
For this specific hook, i think we don't need it. Will remove

mzparacha replied:
On second thoughts, i think we should update state when a user is banned. But it seems that the ban API doesn't return anything. Will do more digging here.

jnaviask replied:
A banned user cannot interact with content in the community (i.e. create threads/comments/reactions). @mzparacha

mzparacha commented on packages/commonwealth/client/scripts/state/api/profiles/getProfilesByAddress.ts:
```diff
@@ -0,0 +1,89 @@
+  profileAddresses: string[];
+}
+
+const fetchProfilesByAddress = async ({
```
indeed yes.

mzparacha commented on packages/commonwealth/client/scripts/state/api/profiles/getProfilesByAddress.ts:
```diff
@@ -0,0 +1,89 @@
+      ApiEndpoints.FETCH_PROFILES,
+      // we should not cache by current chain as it can break logic for DISCOURAGED_NONREACTIVE_getProfilesByAddress
+      // currentChainId,
+      ...profileChainIds,
+      ...profileAddresses,
```
hm, nice spot, we should probably sort the array, maybe by ascending

mzparacha replied:
Done. I went with the sort method.

mzparacha commented on packages/commonwealth/client/scripts/hooks/useWallets.tsx:
```diff
@@ -92,6 +94,8 @@ const useWallets = (walletProps: IuseWalletProps) => {
     onAction: true,
   });
 
+  const { mutateAsync: updateProfile } = useUpdateProfileByAddressMutation({});
```
Done

mzparacha commented on packages/commonwealth/client/scripts/state/api/profiles/getProfilesByAddress.ts:
```diff
@@ -0,0 +1,89 @@
+    // eslint-disable-next-line @tanstack/query/exhaustive-deps
+    queryKey: [
+      ApiEndpoints.FETCH_PROFILES,
+      // we should not cache by current chain as it can break logic for DISCOURAGED_NONREACTIVE_getProfilesByAddress
```
Done

mzparacha commented on packages/commonwealth/client/scripts/state/api/profiles/getProfilesByAddress.ts:
```diff
@@ -0,0 +1,89 @@
+interface UseFetchProfilesByAddressQuery extends FetchProfilesByAddressProps {
+  apiCallEnabled?: boolean;
+}
+const useFetchProfilesByAddressQuery = ({
```
Fixed

mzparacha commented:
@masvelio Really good feedback! Thank you.

I have addressed some of your comments above, and added the 'Done' comment on some places. If the new changes for those are fine, then please feel free to 'Resolve Conversation'. 

I will look into the other feedback comments soon.

masvelio reviewed (commented):
Went through all files.

Re this 
> The newProfile is not entirely cleaned up. I left some event emitters there intentionally as to not break the wallet/auth logic. We need to come up with plan to cleanup manual event emitters. @jnaviask @masvelio 

The fundamental problem I see here is that `User.tsx` component should be a dummy component. This means that it should be only responsible for showing the UI based on the props. However, user component used to make API call and now it still does it but using react-query instead of regular axios call. 

At this point I am not sure if we are able to solve it easily (we are talking about removing event emitter). To make it work in a reactive way probably two things should happen:
1. Untangle the mess around Account, AddressInfo, NewProfile, MinimumProfile. Identify dependencies, make it simpler and move it to the react-query.
2. Make `User.tsx` dummy component. This means that the data should be available in the parent component before the render of User component. Let's take a look on members page.

jnaviask commented:
cc @zakhap to review Marcin's comment above

mzparacha commented on packages/commonwealth/client/scripts/state/api/profiles/getProfilesByAddress.ts:
```diff
@@ -0,0 +1,89 @@
+
+export const DISCOURAGED_NONREACTIVE_getProfilesByAddress = (
+  chainId,
+  address
```
Done

mzparacha commented on packages/commonwealth/client/scripts/state/api/profiles/updateProfileByAddress.ts:
```diff
@@ -0,0 +1,82 @@
+  socials,
+}: UpdateProfileByAddressProps) => {
+  // TODO: ideally this should return a response
+  await axios.post(`${app.serverUrl()}/updateProfile/v2`, {
```
Ticket https://github.com/hicommonwealth/commonwealth/issues/4939

mzparacha replied:
I am implementing the PR changes https://github.com/hicommonwealth/commonwealth/pull/4940

cc: @jnaviask

mzparacha replied:
Done

mzparacha commented on packages/commonwealth/client/scripts/state/api/profiles/updateProfileByAddress.ts:
```diff
@@ -0,0 +1,82 @@
+  avatarUrl,
+  socials,
+}: UpdateProfileByAddressProps) => {
+  // TODO: ideally this should return a response
```
Discussing [here](https://github.com/hicommonwealth/commonwealth/pull/4865#discussion_r1310396048)

mzparacha commented on packages/commonwealth/client/scripts/state/api/profiles/fetchProfilesByAddress.ts:
```diff
@@ -0,0 +1,89 @@
+  );
+};
+
+export default useFetchProfilesByAddressesQuery;
```
It can be `byAddresses` as well.

mzparacha commented on packages/commonwealth/client/scripts/views/components/edit_profile.tsx:
```diff
@@ -151,13 +128,22 @@ const EditProfileComponent = () => {
+      updateProfile({ ...profileUpdate, profileId: profile.id })
+        .then(() => {
+          // manually trigger a redraw
+          NewProfilesController.Instance.isFetched.emit('redraw');
```
Hm, it was being done under the hook by old code, so I kind of migrated it in new code. But you are right, this emit calll is not needed

mzparacha replied:
This was removed. Marking as resolved

mzparacha commented on packages/commonwealth/client/scripts/views/components/user/user.tsx:
```diff
@@ -377,14 +226,14 @@ export const User = ({
   );
 
-  return popover ? (
+  return shouldShowPopover ? (
```
I agree, thanks for the suggestion. Will do this soon.

mzparacha replied:
On second thoughts, I think the scope of file refactors is a little different. To keep things moving I think we can create a new followup ticket for this.

jnaviask replied:
@mzparacha please create a ticket for the changes you scoped out above

mzparacha replied:
Ticket [here](https://github.com/hicommonwealth/commonwealth/issues/4995) @jnaviask. I self assigned.

mzparacha commented:
Addressed some more comments, Thanks for the feedback @masvelio. Will address the remaining comments soon.

mzparacha commented on packages/commonwealth/client/scripts/views/components/proposals/vote_listing.tsx:
```diff
@@ -74,6 +74,17 @@ export const VoteListing = (props: VoteListingProps) => {
     );
   }
 
+  const VoterInfo = ({ voter, shouldShowPopover = true }) => {
+    return (
+      <User
+        userAddress={voter.address}
+        userChainId={voter.chain?.id || voter.profile?.chain}
+        shouldLinkProfile
+        shouldShowPopover={shouldShowPopover}
+      />
+    );
+  };
```
The abstraction was only for some cleanup.

> Or instead of creating new component you can do some func like

Yes we could, buts in code its used sequentially like this (notice the shouldShowPopover prop)

```
<VoterInfo voter={vote.account} />
<VoterInfo voter={vote.account} shouldShowPopover={false} />
<VoterInfo voter={vote.account} shouldShowPopover={false} />
<VoterInfo voter={vote.account} />
```

masvelio commented:
From my perspective code looks good. Would be nice though to have some feedback on my comment, just to have a brief vision for the near future @zakhap https://github.com/hicommonwealth/commonwealth/pull/4865#pullrequestreview-1604479147 
cc @jnaviask

mzparacha commented:
Thanks again for the feedback @masvelio. I will fix the merge conflicts in some time.

jnaviask commented:
@masvelio I think the changes you propose in your comment about the domain model would be fantastic -- and a lot of it isn't just on the domain modeling side, but on the engineering side, of "doing it right" wrt users. The hacks have accumulated over the years and I personally would love for us to cut them out. That said, the difficult part is putting together a plan that gets us from where we are now to where we want to land (+ allocating the time for it).

jnaviask commented:
@mzparacha please fix CI issues (code quality, unit / integration tests)

mzparacha commented:
E2E test seem to be failing due to timeout 
<img width="879" alt="image" src="https://github.com/hicommonwealth/commonwealth/assets/51641047/14e27563-7960-409d-a7a1-fc10d0e3638e">

jnaviask commented on packages/commonwealth/client/scripts/views/components/Comments/CreateComment.tsx:
```diff
@@ -171,7 +171,12 @@ export const CreateComment = ({
           >
-            <User user={author} hideAvatar linkify />
+            <User
+              userAddress={author.address}
```
This change is the cause of the error in the e2e test -- author is undefined, probably because it loads the component without the proper value of `app.user.activeAccount`. Please fix if possible!

Also if you search for all uses of
```
  const author = app.user.activeAccount;
```
You will see several other pages where this may be a problem.

mzparacha replied:
@jnaviask I am not getting undefined author. Tried with these steps

---
#### Secnario # 1

1. Login and go to a joined community (I tried with dydx)
2. Open a thread
3. Add a comment

-- 

#### Secnario # 1

1. Go to a joined community, and open a thread (I tried with dydx)
2. Logout and then login
3. Add a comment

--

> This change is the cause of the error in the e2e test 

Hmm, I am not sure about this, most of them seem to be sequelize errors?

jnaviask reviewed (changes requested):
One issue causing trouble in the e2e tests then should be ready to go.

mzparacha commented:
Looking into it

jnaviask commented:
cc @mzparacha to resolve merge conflicts then this PR is g2g
